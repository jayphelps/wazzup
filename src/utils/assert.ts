export const assert = (value: boolean, msg = 'Assertion failed') => {
  if (value === false) {
    throw new Error(msg);
  }
};
