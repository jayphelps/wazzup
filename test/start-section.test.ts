import { ast, wasm } from '../';
import { assertModule } from './utils/assertModule';

describe('Start Section', () => {
  let module;

  beforeEach(() => {
    module = new ast.Module();
  });

  test('adds a start', () => {
    const type1 = module.addFunctionType();
    const func1 = module.addFunction('foo', type1, [], []);

    module.addStart(func1);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (func (;0;) (type 0))
      |   (start 0))
    `);

    const type2 = module.addFunctionType();
    const func2 = module.addFunction('bar', type2, [], []);

    // overrides the first start (highlander)
    module.addStart(func2);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (type (;1;) (func))
      |   (func (;0;) (type 0))
      |   (func (;1;) (type 1))
      |   (start 1))
    `);
  });
});
