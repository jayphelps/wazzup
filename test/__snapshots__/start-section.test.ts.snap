// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Start Section adds a start 1`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "start": Function {
    "body": Array [],
    "externalKind": 0,
    "functionType": FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
    "localTypes": Array [],
    "name": "foo",
    "type": 112,
  },
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Start Section adds a start 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  4,
  1,
  96,
  0,
  0,
  3,
  2,
  1,
  0,
  8,
  1,
  0,
  10,
  4,
  1,
  2,
  0,
  11,
]
`;

exports[`Start Section adds a start 3`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "bar",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "start": Function {
    "body": Array [],
    "externalKind": 0,
    "functionType": FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
    "localTypes": Array [],
    "name": "bar",
    "type": 112,
  },
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Start Section adds a start 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  7,
  2,
  96,
  0,
  0,
  96,
  0,
  0,
  3,
  3,
  2,
  0,
  1,
  8,
  1,
  1,
  10,
  7,
  2,
  2,
  0,
  11,
  2,
  0,
  11,
]
`;
